<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Process genotype data from survival experiment</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Process genotype data from survival experiment
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;rsurvival&rsquo; version 0.0.0.9000</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="AnalyseExpt.html">AnalyseExpt</a></td>
<td>A function to analyse a selection experiment, including observed frequencies of genotype and probabilities that they are due to random picking and positive selection.</td></tr>
<tr><td style="width: 25%;"><a href="AnalyseSplittedExpt.html">AnalyseSplittedExpt</a></td>
<td>A function to analyse a selection experiment, including observed frequencies of genotype and probabilities that they are due to random picking and positive selection.</td></tr>
<tr><td style="width: 25%;"><a href="CalcFreqGt.html">CalcFreqGt</a></td>
<td>A function to count occurences of genotypes in a genotype matrix.</td></tr>
<tr><td style="width: 25%;"><a href="CalcFreqVariant.html">CalcFreqVariant</a></td>
<td>A function to count occurences of genotypes in a variant vector.</td></tr>
<tr><td style="width: 25%;"><a href="CalcOddedPick.html">CalcOddedPick</a></td>
<td>A function to calculate the number of picked lives per categories of samples according to MWNCHypergeo model.</td></tr>
<tr><td style="width: 25%;"><a href="CalcProbsSelection.html">CalcProbsSelection</a></td>
<td>A function to calculate probabilities that observed frequencies are due to positive selection using MWNCHHypergeo model.</td></tr>
<tr><td style="width: 25%;"><a href="CalcWeightsPredation.html">CalcWeightsPredation</a></td>
<td>A function to calculate odds for predation from observed selection frequencies using MWNCHHypergeo model.</td></tr>
<tr><td style="width: 25%;"><a href="CensorLowDepth.html">CensorLowDepth</a></td>
<td>A function to replace genotype data with low depth by NA</td></tr>
<tr><td style="width: 25%;"><a href="ConvertFileToLogic.html">ConvertFileToLogic</a></td>
<td>A function to turn a file into a logical vector. Useful for making a survivals vector to split your data in two parts.</td></tr>
<tr><td style="width: 25%;"><a href="CountSignSnps.html">CountSignSnps</a></td>
<td>A function to count snps with a low probability of being due to a neutral model</td></tr>
<tr><td style="width: 25%;"><a href="EstimateFST.html">EstimateFST</a></td>
<td>A function to estimate the FST of a variant using the Weir &amp; Cockerham (1984) method</td></tr>
<tr><td style="width: 25%;"><a href="ExtractGt.html">ExtractGt</a></td>
<td>A function to extract a genotype from a vcfR::VCF object.</td></tr>
<tr><td style="width: 25%;"><a href="FindIdsAlFreqs.html">FindIdsAlFreqs</a></td>
<td>A function to find the ids of freq.al columns in a frequency matrix.</td></tr>
<tr><td style="width: 25%;"><a href="FindIdsGtCounts.html">FindIdsGtCounts</a></td>
<td>A function to find the ids of counts.gt columns in a frequency matrix.</td></tr>
<tr><td style="width: 25%;"><a href="IsAtLeastNARatio.html">IsAtLeastNARatio</a></td>
<td>A function to check if the ratio of NA in a vector is over a threshold.</td></tr>
<tr><td style="width: 25%;"><a href="IterRandPick.html">IterRandPick</a></td>
<td>A function to iterate significant snps counting in a genotype matrx with a random permutation of the survival vector</td></tr>
<tr><td style="width: 25%;"><a href="LoadVcf.html">LoadVcf</a></td>
<td>A function to get a filtered (and possibly splitted) vcfR::VCF object from a VCF file.</td></tr>
<tr><td style="width: 25%;"><a href="OmitPoorSamples.html">OmitPoorSamples</a></td>
<td>A function to remove columns with a high ratio of missing data from a dataframe. (e.g. samples from genotype)</td></tr>
<tr><td style="width: 25%;"><a href="OmitPoorVariants.html">OmitPoorVariants</a></td>
<td>A function to remove rows with a high ratio of missing data from a dataframe. (e.g. variants from genotype)</td></tr>
<tr><td style="width: 25%;"><a href="PermutRandSurv.html">PermutRandSurv</a></td>
<td>A function to calculate probabilities of neutral model and positive selection model from a genotype matrix whit a randomly shuffled survival vector</td></tr>
<tr><td style="width: 25%;"><a href="RmIndels.html">RmIndels</a></td>
<td>A function to remove indel variants from a vcfR::VCF object</td></tr>
<tr><td style="width: 25%;"><a href="RmNonBiallelics.html">RmNonBiallelics</a></td>
<td>A function to remove non-biallelic variants from a vcfR::VCF object</td></tr>
<tr><td style="width: 25%;"><a href="SliceDfColumns.html">SliceDfColumns</a></td>
<td>A function to slice data frames without loosing column names</td></tr>
<tr><td style="width: 25%;"><a href="SliceDfRows.html">SliceDfRows</a></td>
<td>A function to slice data frames without loosing row names</td></tr>
<tr><td style="width: 25%;"><a href="SplitGt.html">SplitGt</a></td>
<td>A function to split a genotype matrix two genotype matrix according to a logical vector</td></tr>
<tr><td style="width: 25%;"><a href="SplitVcf.html">SplitVcf</a></td>
<td>A function to split a vcfR::VCF object in two vcfR::VCF object according to a logical vector</td></tr>
<tr><td style="width: 25%;"><a href="SplitVect.html">SplitVect</a></td>
<td>A function to split a vector into two vectors according to a survival logical vector</td></tr>
</table>
</body></html>
